### PR TITLE
Add LookupBoundary to Material

### PR DESCRIPTION
**Pull request by goderbauer**
Material.of and Material.maybeOf (as well as debugCheckHasMaterial) cannot find Materials that are past a lookup boundary.

### CODE REVIEW

File: packages/flutter/lib/src/widgets/lookup_boundary.dart [hunk 1]
```diff
@@ -250,6 +250,53 @@ class LookupBoundary extends InheritedWidget {
...
+  /// Returns true if a [LookupBoundary] is hiding the nearest
+  /// [Widget] of the specified type `T` from the provided [BuildContext].
+  ///
+  /// This method throws when asserts are disabled.
+  static bool debugIsHidingAncestorWidgetOfExactType<T extends Widget>(BuildContext context) {
+    bool? result;
+    assert(() {
+      bool hiddenByBoundary = false;
+      bool ancestorFound = false;
+      context.visitAncestorElements((Element ancestor) {
```
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/lookup_boundary.dart:262]: Oh `LookupBoundary.maybeOf(context).` registers `context` as a dependant I guess?

**Review by LongCatIsLooong — APPROVED** [review]

File: packages/flutter/lib/src/material/debug.dart [hunk 2]
```diff
@@ -32,12 +33,17 @@ import 'scaffold.dart' show Scaffold, ScaffoldMessenger;
...
   assert(() {
-    if (context.widget is! Material && context.findAncestorWidgetOfExactType<Material>() == null) {
+    if (LookupBoundary.findAncestorWidgetOfExactType<Material>(context) == null) {
+      final bool hiddenByBoundary = LookupBoundary.debugIsHidingAncestorWidgetOfExactType<Material>(context);
       throw FlutterError.fromParts(<DiagnosticsNode>[
-        ErrorSummary('No Material widget found.'),
+        ErrorSummary('No Material widget found${hiddenByBoundary ? ' within the closest LookupBoundary' : ''}.'),
+        if (hiddenByBoundary)
+          ErrorDescription(
+            'There is an ancestor Material widget, but it is hidden by a LookupBoundary.'
```
[the remark this answers is not in the document]
Comment (goderbauer) [reply packages/flutter/lib/src/material/debug.dart:42]: Not a bad idea, do you man the creator chain returned by https://api.flutter.dev/flutter/widgets/Element/debugGetCreatorChain.html or did you have something else in mind?

[the remark this answers is not in the document]
Comment (goderbauer) [reply packages/flutter/lib/src/material/debug.dart:42]: Hm, one problem is that there could be potentially multiple LookupBoundaries blocken access to the element in question. The error message could get really complex if we show all of their chains…

File: packages/flutter/lib/src/widgets/lookup_boundary.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (goderbauer) [reply packages/flutter/lib/src/widgets/lookup_boundary.dart:262]: Those methods also return Widgets, which don’t have ‘depth’ getters. :)

> also it's a bit strange that there are 3 potential states (no material in the tree, no material within the LUB, material found) but we're just returning a boolean.

That’s why I phrased this method as “is there a material widget that’s hidden from the BuildContext by a LUB”. The answer to that question is false for the “material found” and “no material” case. I don’t see any use case w.r.t. LUB that requires differentiating between these states?

File: packages/flutter/lib/src/material/debug.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/material/debug.dart:42]: maybe a debugLabel then?

[the remark this answers is not in the document]
Comment (goderbauer) [reply packages/flutter/lib/src/material/debug.dart:42]: Talked offline, we will reconsider this once we actually have more LUB in use to see what information would be most useful here.

**Pull request merged**